src/asymmetric_encryption.cpp

```diff
@@ -0,0 +1,231 @@
+                                             static_cast<int>(oaepPaddingMode.getLabel().size()));
+
+                /* Release ownership from the unique_ptr since the function above takes ownership of
+                 * the label pointer unless it throws an exception*/ 
+                label_copy.release();
+            }
+        }
+
+        /* First call to determine the buffer length */
+        _EVP_PKEY_encrypt(keyCtx.get(),
+                          nullptr,
+                          &encryptedMessageLen,
+                          reinterpret_cast<const unsigned char *>(&message.getByteArray()[0]),
+                          message.getByteArray().size());
+
+        /* Allocate memory for the buffer, based on the size returned by _EVP_PKEY_encrypt */
+        encryptedMessage.reset(static_cast<unsigned char*>(_OPENSSL_malloc(static_cast<int>(encryptedMessageLen))));
+
+        /* Second call to perform the actual encryption */
+        _EVP_PKEY_encrypt(keyCtx.get(),
+                          encryptedMessage.get(),
+                          &encryptedMessageLen,
+                          reinterpret_cast<const unsigned char *>(&message.getByteArray()[0]),
+                          message.getByteArray().size());
+
+    } catch (const OpenSSLException &e) {
+        throw MoCOCrWException(e.what());
+    }
+
+    return std::vector<uint8_t>(encryptedMessage.get(), encryptedMessage.get() + encryptedMessageLen);
+}
+
+AsymmetricEncryption::CryptoData AsymmetricEncryption::decrypt(AsymmetricPrivateKey key,
+                                                               const RSAPadding &pad,
+                                                               const CryptoData& message)
+{
+    size_t decryptedMessageLen{0};
+    auto keyCtx = _EVP_PKEY_CTX_new(key.internal());
+    SSL_RSA_OAEP_LABEL_Ptr label_copy(nullptr);
+    SSL_RSA_ENCRYPTION_DATA_Ptr decryptedMessage{nullptr};
+
+    try {
+        const auto paddingMode = pad.getPadding();
+
+        _CRYPTO_malloc_init();
+
+        _EVP_PKEY_decrypt_init(keyCtx.get());
+        _EVP_PKEY_CTX_set_rsa_padding(keyCtx.get(), static_cast<int>(paddingMode));
+
+        if(openssl::RSAPaddingMode::OAEP == paddingMode) {
+            const auto oaepPaddingMode = static_cast<const OAEPPadding&>(pad);
+
+            _EVP_PKEY_CTX_set_rsa_oaep_md(keyCtx.get(),
+                                          _getMDPtrFromDigestType(oaepPaddingMode.getHashingFunction()));
+
+            _EVP_PKEY_CTX_set_rsa_mgf1_md(keyCtx.get(),
+                                          _getMDPtrFromDigestType(oaepPaddingMode.getMaskingFunction()));
+
+            if(oaepPaddingMode.getLabel().size() > 0){
+                
+                /* Make a copy of the label, since the context takes ownership of it when calling
+                 * '_EVP_PKEY_CTX_set_rsa_oaep_label()' function */
+                label_copy.reset(static_cast<uint8_t*>(_OPENSSL_malloc(oaepPaddingMode.getLabel().size())));
```

Comment:
An issue due to a implicit conversion which losses precision may lead to a severe issue regarding to memory allocation and memory access.  

1. _OPENSSL_malloc receives an int, which is obtained from a implicit conversion of unsigned long to int and may result in negative number. And therefore, try to allocated negative memory. 

2. The value used for the memory allocation by the function OPENSSL_malloc may be different (due to the implicit conversion) from the value use in the memcpy which defined the amount o memory allocated/copied.

I want to highlight, that such case would only appends if  "oaepPaddingMode" size exceeds the int max value and may not be possible to use such padding Mode. 

I would like to ear your option about this topic. However, i suggest that at least a comment to be added at function header to warning for such cases.
Please take in consideration the issue and advise

Reply:
Same as above